core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -144,40 +144,131 @@ namespace kagome::consensus::babe {
   }
 
   bool BabeImpl::prepare() {
-    auto res = getInitialEpochDescriptor();
-    if (res.has_error()) {
-      SL_CRITICAL(log_, "Can't get initial epoch descriptor: {}", res.error());
+    auto initial_epoch_res = getInitialEpochDescriptor();
+    if (initial_epoch_res.has_error()) {
+      SL_CRITICAL(log_,
+                  "Can't get initial epoch descriptor: {}",
+                  initial_epoch_res.error());
       return false;
     }
 
     best_block_ = block_tree_->bestLeaf();
 
-    // Check if best block has state for usual sync method
-    if (app_config_.syncMethod() == SyncMethod::Full) {
-      auto best_block_header_res =
-          block_tree_->getBlockHeader(best_block_.hash);
-      if (best_block_header_res.has_error()) {
-        SL_CRITICAL(log_,
-                    "Can't get header of best block ({}): {}",
-                    best_block_,
-                    best_block_header_res.error());
-        return false;
-      }
-      const auto &best_block_header = best_block_header_res.value();
+    auto best_block_header_res = block_tree_->getBlockHeader(best_block_.hash);
+    if (best_block_header_res.has_error()) {
+      SL_CRITICAL(log_,
+                  "Can't get header of best block ({}): {}",
+                  best_block_,
+                  best_block_header_res.error());
+      return false;
+    }
+    const auto &best_block_header = best_block_header_res.value();
+    const auto &state_root = best_block_header.state_root;
+
+    std::chrono::microseconds warp_sync_duration;
+    std::chrono::microseconds fast_sync_duration;
+    std::chrono::microseconds full_sync_duration;
+
+    // Calculate lag our best block by slots
+    BabeSlotNumber lag_slots = 0;
+    if (auto babe_digests_res = getBabeDigests(best_block_header);
+        babe_digests_res.has_value()) {
+      auto &[seal, babe_header] = babe_digests_res.value();
+      lag_slots = babe_util_->getCurrentSlot() - babe_header.slot_number;
+    }
+
+    // WARP: n * header_loading / k + state_loading + lag * block_execution
+    //       {               catchup              }
+    // FAST: n * header_loading + state_loading + lag' * block_execution
+    //       {             catchup'           }
+    // FULL: n * block_execution + lag" * block_execution
+    //       {     catchup"    }
+
+    {
```

Comment:
Can we move calculating durations into separate functions?

Reply:
Sure. Done